modules/crypt.cpp

```diff
@@ -43,16 +42,151 @@
+    /* Generate our keys once and reuse, just like ssh keys */
+    DH *dh = NULL;
+    CString sPrivKey;
+    CString sPubKey;
+
+#if OPENSSL_VERSION_NUMBER < 0X10100000L
+    int DH_set0_pqg(DH *dh, BIGNUM *p, BIGNUM *q, BIGNUM *g) {
+        /* If the fields p and g in dh are NULL, the corresponding input
+         * parameters MUST be non-NULL.  q may remain NULL.
+         */
+        if (dh == NULL || (dh->p == NULL && p == NULL) || (dh->g == NULL && g == NULL))
+            return 0;
+
+        if (p != NULL) {
+            BN_free(dh->p);
+            dh->p = p;
+        }
+        if (g != NULL) {
+            BN_free(dh->g);
+            dh->g = g;
+        }
+        if (q != NULL) {
+            BN_free(dh->q);
+            dh->q = q;
+            dh->length = BN_num_bits(q);
+        }
+
+        return 1;
+    }
+
+    void DH_get0_key(const DH *dh, const BIGNUM **pub_key, const BIGNUM **priv_key) {
+        if (dh != NULL) {
+            if (pub_key != NULL)
+                *pub_key = dh->pub_key;
+            if (priv_key != NULL)
+                *priv_key = dh->priv_key;
+        }
+    }
+
+#endif
+
+    bool DH1080_gen() {
+        /* Generate our keys on first call */
+        if (!dh) {
+            int len;
+            unsigned char *buf;
+            const BIGNUM *bPrivKey = NULL;
+            const BIGNUM *bPubKey = NULL;
+            BIGNUM *bPrime = NULL;
+            BIGNUM *bGen = NULL;
+            dh = DH_new();
+
+            if (!BN_hex2bn(&bPrime, prime1080) || !BN_dec2bn(&bGen, "2") || !DH_set0_pqg(dh, bPrime, NULL, bGen) || !DH_generate_key(dh)) {
+                /* one of them failed */
+                if (bPrime)
+                    BN_clear_free(bPrime);
+                if (bGen)
+                    BN_clear_free(bGen);
+                if (dh) {
+                    DH_free(dh);
+                    dh = NULL;
+                }
+                return false;
+            }
+
+            /* Get our keys */
+            DH_get0_key(dh, &bPubKey, &bPrivKey);
+
+            /* Get our private key */
+            len = BN_num_bytes(bPrivKey);
+            buf = (unsigned char *)calloc(len, 1);
+            BN_bn2bin(bPrivKey, buf);
+            /*
+             * Use basic_string constructor with given start and end,
+             *  as we might have \0 and don't want to cast
+             */
+            sPrivKey = CString::basic_string(buf, &buf[len]);
+            sPrivKey.Base64Encode();
+
+            /* Get our public key */
+            len = BN_num_bytes(bPubKey);
+            buf = (unsigned char *)realloc(buf, len);
+            BN_bn2bin(bPubKey, buf);
+            sPubKey = CString::basic_string(buf, &buf[len]);
+            sPubKey.Base64Encode();
```

Comment:
Not that it matters much, but you could just `string.resize(len)` a (previously empty) string to the right size and then call `BN_bn2bin(foo, string.c_str())` to directly write the binary representation there. That would save a copy (not important and all) and would avoid manual memory management.